src/assisted_installer_controller/assisted_installer_controller.go

```diff
@@ -247,6 +247,168 @@ func (c controller) UpdateBMHs(wg *sync.WaitGroup) {
 	}
 }
 
+func (c controller) UpdateWorkerLinkage(wg *sync.WaitGroup) {
+	defer wg.Done()
+	log := c.log.WithField("Job", "WorkerLinker")
+	for {
+		time.Sleep(GeneralWaitInterval)
+
+		exists, err := c.kc.IsMetalProvisioningExists()
+		if err != nil {
+			continue
+		}
+		if err == nil && exists {
+			log.Infof("Provisioning CR exists, no need to update BMHs")
+			return
+		}
+
+		machines, err := c.kc.ListMachines()
+		if err != nil {
+			log.WithError(err).Errorf("Failed to get nodes")
+			continue
+		}
+
+		nodes, err := c.kc.ListNodes()
+		if err != nil {
+			log.WithError(err).Errorf("Failed to get nodes")
+			continue
+		}
+
+		bmhs, err := c.kc.ListBMHs()
+		if err != nil {
+			log.WithError(err).Errorf("Failed to BMH hosts")
+			continue
+		}
+
+		allUpdated := c.updateWorkerLinkage(bmhs, nodes, machines)
+		if allUpdated {
+			log.Infof("Updated all the worker links, finished successfully")
+			return
+		}
+	}
+}
+
+func (c controller) updateWorkerLinkage(bmhList metal3v1alpha1.BareMetalHostList, nodeList *v1.NodeList, machineList *mapiv1beta1.MachineList) bool {
```

Comment:
a question regarding the flow: why do we need to find the node? Why not simply pick BMH without ConsumerRef, and then pick an available machine?

Reply:
OK, i'l give that ago. I was trying to set the internalIP and I needed the node. but now that i am not doing that, i can remove the node stuff.